daemon/api_themes.go

```diff
@@ -0,0 +1,249 @@
+		// Skip duplicates
+		if _, ok := status[theme]; ok {
+			continue
+		}
+		if strutil.SortedListContains(installed, theme) {
+			status[theme] = themeInstalled
+			continue
+		}
+		status[theme] = themeUnavailable
+		for _, name := range themePackageCandidates(prefix, theme) {
+			var info *snap.Info
+			if info, err = theStore.SnapInfo(ctx, store.SnapSpec{Name: name}, user); err == store.ErrSnapNotFound {
+				continue
+			} else if err != nil {
+				return nil, err
+			}
+			// Only mark the theme as available if it has
+			// been published to the stable channel.
+			if info.Channel == "stable" {
+				status[theme] = themeAvailable
+				toInstall[name] = true
+			}
+			break
+		}
+	}
+	return status, nil
+}
+
+func themeStatusAndMissingSnaps(ctx context.Context, c *Command, user *auth.UserState, gtkThemes, iconThemes, soundThemes []string) (status themeStatusResponse, err error) {
+	installedGtk, installedIcon, installedSound, err := installedThemes(c)
+	if err != nil {
+		return themeStatusResponse{}, err
+	}
+
+	theStore := getStore(c)
+	candidates := make(map[string]bool)
+	if status.GtkThemes, err = themeStatusForPrefix(ctx, theStore, user, "gtk-theme-", gtkThemes, installedGtk, candidates); err != nil {
+		return themeStatusResponse{}, err
+	}
+	if status.IconThemes, err = themeStatusForPrefix(ctx, theStore, user, "icon-theme-", iconThemes, installedIcon, candidates); err != nil {
+		return themeStatusResponse{}, err
+	}
+	if status.SoundThemes, err = themeStatusForPrefix(ctx, theStore, user, "sound-theme-", soundThemes, installedSound, candidates); err != nil {
+		return themeStatusResponse{}, err
+	}
+	status.MissingSnaps = make([]string, 0, len(candidates))
+	for pkg := range candidates {
+		status.MissingSnaps = append(status.MissingSnaps, pkg)
+	}
+	sort.Strings(status.MissingSnaps)
+
+	return status, nil
+}
+
+func checkThemes(c *Command, r *http.Request, user *auth.UserState) Response {
+	ctx := store.WithClientUserAgent(r.Context(), r)
+	q := r.URL.Query()
+	status, err := themeStatusAndMissingSnaps(ctx, c, user, q["gtk-theme"], q["icon-theme"], q["sound-theme"])
+	if err != nil {
+		return InternalError("cannot get theme status: %s", err)
+	}
+
+	return SyncResponse(status, nil)
+}
+
+type themeInstallReq struct {
+	GtkThemes   []string `json:"gtk-themes"`
+	IconThemes  []string `json:"icon-themes"`
+	SoundThemes []string `json:"sound-themes"`
+}
+
+func installThemes(c *Command, r *http.Request, user *auth.UserState) Response {
+	decoder := json.NewDecoder(r.Body)
+	var req themeInstallReq
+	if err := decoder.Decode(&req); err != nil {
+		return BadRequest("cannot decode request body: %v", err)
+	}
+
+	ctx := store.WithClientUserAgent(r.Context(), r)
+	status, err := themeStatusAndMissingSnaps(ctx, c, user, req.GtkThemes, req.IconThemes, req.SoundThemes)
+	if err != nil {
+		return InternalError("cannot get theme status: %s", err)
+	}
+
+	if len(status.MissingSnaps) == 0 {
+		return BadRequest("no snaps to install")
```

Comment:
why make this an error?

Reply:
My original thought was that this API would only be called if a GET request had been made before hand that reported that candidate snaps were available.  Of course, there could be some time between the two calls that makes that information out of date.

It's not clear we'd want this API to sometimes return a SyncResponse for the "no snaps to install" case and AsyncResponse when there is work to do though.  I couldn't see anything similar to that in other endpoints.  Maybe returning an error with an identifiable error kind would make sense so the client can distinguish it from others?  Maybe reusing `ErrKindSnapAlreadyInstalled` would make sense?

Reply:
no, I double checked, there is precedent for this,  install many also returns a BadRequest if zero snaps are passed in